src/coreclr/jit/lower.cpp

```diff
@@ -8003,6 +8011,42 @@ void Lowering::LowerStoreIndirCoalescing(GenTreeStoreInd* ind)
             return;
         }
 
+        // Now the hardest part: decide whether it's safe to use an unaligned write.
+        //
+        // IND<byte> is always fine (and all IND<X> created here from such)
+        const bool knownUnaligned =
+            ((ind->gtFlags & GTF_IND_KNOWN_UNALIGNED) != 0) && ((prevInd->gtFlags & GTF_IND_KNOWN_UNALIGNED) != 0);
+        if (!knownUnaligned && (genTypeSize(ind) > 1))
+        {
+            // TODO-CQ: if we see that the target is a local memory (not address exposed)
+            // we can use any type (including SIMD) for a new load.
+
+            // Ignore indices for now, they can invalidate our alignment assumptions.
+            if (currData.index != nullptr)
+            {
+                return;
+            }
+
+            // If the baseAddr is not TYP_REF we likely have no hints about its alignment.
+            // while TYP_REF guarantees pointer-size alignment.
+            if (!currData.baseAddr->TypeIs(TYP_REF))
+            {
+                return;
+            }
+
+#ifdef TARGET_XARCH
+            // Check whether data can be guaranteed to be in a cache line.
+            // It seems that only x86 can guarantee that.
+            int startOffset = min(prevData.offset, currData.offset) % genTypeSize(ind);
+            if ((startOffset + (genTypeSize(ind) * 2)) > TARGET_POINTER_SIZE)
+            {
+                return;
+            }
+#else
+            return;
+#endif
+        }
+
```

Comment:
@SingleAccretion @jakobbotsch addressed the concerns about atomicity for primitives we discussed offline.